### PR TITLE
feat(ignoreFiles): added an ignore function to allow granular file exclusions

### PR DESCRIPTION
**Pull request by ivansieder**
Refers to #294

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "copy-webpack-plugin",
-  "version": "4.5.2",
+  "version": "4.6.0",
```
Comment (evilebottnawi) [inline package.json]: Revert this, changelog and lock file - this do contributors

File: src/index.js [hunk 2]
```diff
@@ -92,17 +93,17 @@ function CopyWebpackPlugin(patterns = [], options = {}) {
...
 
             Promise.all(tasks)
-            .catch((err) => {
-                compilation.errors.push(err);
-            })
-            .then(() => callback());
+                .catch((err) => {
+                    compilation.errors.push(err);
+                })
+                .then(() => callback());
```
Comment (evilebottnawi) [inline src/index.js]: Don't modify code style in this PR, this will be do in next update code based

File: src/preProcessPattern.js [hunk 3]
```diff
@@ -60,43 +65,43 @@ export default function preProcessPattern(globalRef, pattern) {
...
+                pattern.glob = escape(pattern.absoluteFrom);
+                pattern.fromArgs = {
+                    dot: true
+                };
+                fileDependencies.push(pattern.absoluteFrom);
+            } else if(!pattern.fromType) {
+                info(`Unrecognized file type for ${pattern.from}`);
+            }
+            return pattern;
+        });
```
Comment (evilebottnawi) [inline src/preProcessPattern.js]: Avoid modify code style

File: src/writeFile.js [hunk 4]
```diff
@@ -11,109 +11,109 @@ export default function writeFile(globalRef, pattern, file) {
...
+                    info(`writing '${file.webpackTo}' to compilation assets from '${file.absoluteFrom}'`);
+                    compilation.assets[file.webpackTo] = {
+                        size: function() {
+                            return stat.size;
+                        },
+                        source: function() {
+                            return content;
+                        }
+                    };
+                });
```
Comment (evilebottnawi) [inline src/writeFile.js]: Avoid modify code based

**Comment by evilebottnawi** [comment]
@ivansieder why don't use `ignore` option?

**Comment by ivansieder** [comment]
Well but that would mean that if I want to exclude single files and a glob pattern, that I have to write custom logic for handling glob patterns, which is pretty difficult IMHO

**Comment by evilebottnawi** [comment]
@ivansieder can you provide example?

**Comment by ivansieder** [comment]
@evilebottnawi an example could be that someone wants to (simplified) ignore all .DS_STORE files which appear somewhere but also wants to exclude explicitly some single index.html files from being copied (that's taken from the Vue CLI example).

**Comment by evilebottnawi** [comment]
@ivansieder what about allow ignore be `String/Function/Array<String|Function>` (i.e. allow be function and allow be function in array)?

**Comment by evilebottnawi** [comment]
@ivansieder sounds reasonable, maybe rename `ignoreFn` to `ignoreFiles` to avoid misleading and in next major release rename `ignore` to `ignoreGlob`

**Comment by ivansieder** [comment]
Done :)

File: src/processPattern.js [hunk 5]
```diff
@@ -73,6 +73,19 @@ export default function processPattern(globalRef, pattern) {
                 }
             }
 
+            // Run ignore functions
+            for (let i = 0; i < pattern._ignoreFunctions.length; i++) {
+                debug(`running ignore function against ${file.relativeFrom}`);
+                if (
+                    pattern._ignoreFunctions[i]({
+                        absolutePath: file.absoluteFrom,
+                        relativePath: file.relativeFrom
```
Comment (evilebottnawi) [inline src/processPattern.js]: I think better arguments is `absolutePath` and `context`, it is allow to know `relative` path and also allow get context

**Comment by evilebottnawi** [comment]
@ivansieder no commits :smile:

File: src/processPattern.js (hunk 5 above)
Comment (ivansieder) [reply src/processPattern.js]: I've named it absolute and relative path, as that's mostly related to absoluteFrom and relativeFrom.

File: src/preProcessPattern.js [hunk 6]
```diff
@@ -48,6 +48,11 @@ export default function preProcessPattern(globalRef, pattern) {
   pattern.context = normalizePath(pattern.context);
   pattern.ignore = globalRef.ignore.concat(pattern.ignore || []);
 
+  // One global and one pattern-specific ignore function is allowed
+  pattern._ignoreFunctions = [];
+  if (globalRef.ignoreFiles && typeof globalRef.ignoreFiles === 'function') pattern._ignoreFunctions.push(globalRef.ignoreFiles);
```
Comment (evilebottnawi) [inline src/preProcessPattern.js]: Avoid `globalRef`, `ignore` and `context` on global level will be removed in next major release in favor `pattern.ignore` and `pattern.context`

File: src/processPattern.js [hunk 7]
```diff
@@ -90,6 +90,21 @@ export default function processPattern(globalRef, pattern) {
             logger.debug(`${glob} doesn't match ${file.relativeFrom}`);
           }
 
+          // Run ignore functions
+          for (let i = 0; i < pattern.allIgnoreFunctions.length; i++) {
```
Comment (evilebottnawi) [inline src/processPattern.js]: Use `forEach`

File: src/preProcessPattern.js (hunk 6 above)
Comment (ivansieder) [reply src/preProcessPattern.js]: so there should be no global pattern possible anymore? just on the pattern directly?

Comment (evilebottnawi) [reply src/preProcessPattern.js]: yes

**Comment by evilebottnawi** [comment]
Also please provide use case where you need `ignore` as glob pattern and `ignoreFiles` as function, because logic is misleading and no need add new option as i say above

**Comment by ivansieder** [comment]
@evilebottnawi would you mind giving it another look?

> 
> 
> Also please provide use case where you need `ignore` as glob pattern and `ignoreFiles` as function, because logic is misleading and no need add new option as i say above

One use case is the Vue.js CLI 3. For multi-page applications, there might be multiple subdirectories, where an index.html needs to be ignored. Using the glob pattern doesn't work in this case, as not all index.html files have to be ignored.

File: src/processPattern.js (hunk 7 above)
Comment (ivansieder) [reply src/processPattern.js]: i've chosen the for loop, as i need to return the resolved promise, if one pattern matches. can I leave it like that for this use case?

**Comment by evilebottnawi** [comment]
@ivansieder Please provide example, I still can not understand why we need a new option when we can implement `ignore` as `Function`

**Comment by ivansieder** [comment]
> 
> 
> @ivansieder Please provide example, I still can not understand why we need a new option when we can implement `ignore` as `Function`

It WOULD actually be possible to use `ignore` as a function, but does that actually make sense? That would mean, that a user, who wants to use the glob pattern needs to evaluate the glob himself. Or how could a design, which combines the glob pattern and a function, look like? What's your idea?

**Comment by evilebottnawi** [comment]
@ivansieder you should install `minimatch` and check this in function, we can't cover all edge cases so we provide function to allow any difficult ignore cases

**Comment by ivansieder** [comment]
@evilebottnawi I'm afraid I can't follow you now anymore. What exactly should/could this function look like, or does that even make sense? Otherwise I'll proceed with closing this PR and report that back to the Vue.js-related issue for now.

**Comment by evilebottnawi** [comment]
we should implement `ignore` as `Function`

**Comment by ivansieder** [comment]
you mean unifying ignore and ignoreFiles? so that way someone would have to do the minimatch on his own and this would introduce a breaking change. is that correct?

**Comment by evilebottnawi** [comment]
@ivansieder yes, if you use `ignore` as `Function` you need own logic for ignore (you can use any libraries `minimatch`, `anymatch`, `RegExp` and etc)

**Comment by ivansieder** [comment]
alright so the proposal would be to let `ignore` be either an array of globs or a function and if it's an array, preserve the old behavior and otherwise use the function?

**Comment by evilebottnawi** [comment]
@ivansieder yes :+1:

**Comment by ivansieder** [comment]
alright, and if the global setting is an array of globs, probably also the pattern-specific ignore value should be an array of globs? or should it be possible to have an array of globs globally and a function as pattern-specific value and vice-versa?

**Comment by evilebottnawi** [comment]
I think we can avoid global option because it will be removed in next major release

**Comment by ivansieder** [comment]
The purpose for the Vue.js case has been solved differently.

**Pull request closed without merging**